universal-application-tool-0.0.1/app/repository/VersionRepository.java

```diff
@@ -304,15 +308,18 @@ public void updateProgramsForNewDraftQuestion(long oldId) {
     return database.find(Version.class).findList();
   }
 
-  public void setLive(long versionId) {
-    Version draftVersion = getDraftVersion();
-    Version activeVersion = getActiveVersion();
+  /** Sets a previous version to ACTIVE again, and hides any DRAFT version. */
+  public void setLiveVersion(long versionId) {
+    Version currentDraftVersion = getDraftVersion();
+    Version currentActiveVersion = getActiveVersion();
     Version newActiveVersion = database.find(Version.class).setId(versionId).findOne();
-    newActiveVersion.setLifecycleStage(LifecycleStage.ACTIVE);
-    newActiveVersion.save();
-    activeVersion.setLifecycleStage(LifecycleStage.OBSOLETE);
-    activeVersion.save();
-    draftVersion.setLifecycleStage(LifecycleStage.DELETED);
-    draftVersion.save();
+    Preconditions.checkState(
+        currentDraftVersion.id != versionId, "Can't rollback to the DRAFT version.");
+    Preconditions.checkState(
+        currentActiveVersion.id != versionId, "Can't rollback to the current ACTIVE version.");
+
+    newActiveVersion.setLifecycleStage(LifecycleStage.ACTIVE).save();
+    currentActiveVersion.setLifecycleStage(LifecycleStage.OBSOLETE).save();
+    currentDraftVersion.setLifecycleStage(LifecycleStage.DELETED).save();
```

Comment:
question (non-blocking): are these three saves done in a transaction? if not they should be

Reply:
Really unclear to me.  I'll figure that out subsequently.  There is a test that seems to imply so, but I have no idea what it's actually doing.  Do you?

https://github.com/seattle-uat/civiform/blob/284d7e3f289e490fdd221341a75e951e9172c00a/universal-application-tool-0.0.1/test/repository/VersionRepositoryTest.java#L80

Reply:
Looking at the code, they are not run in a transaction. Something for us to add and potentially a source of data inconsistency, though if this had gone we'd have seen different problems than the ones we have.

Reply:
Do you know what the test is doing? or trying to do?